File: Python3.x/1038-Binary_Search_Tree_to_Greater_Sum_Tree.py
# Runtime: 36 ms, faster than 73.66% of Python3 online submissions for Binary Search Tree to Greater Sum Tree.
# Memory Usage: 13.3 MB, less than 100.00% of Python3 online submissions for Binary Search Tree to Greater Sum Tree.
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    val = 0
    def bstToGst(self, root: TreeNode) -> TreeNode:
        if root.right:
            self.bstToGst(root.right)
        root.val = self.val = self.val + root.val
        if root.left:
            self.bstToGst(root.left)
        return root

File: Algorithms/C++/1038-Binary_Search_Tree_to_Greater_Sum_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* bstToGst(TreeNode* root) {
        if (root->right)
            bstToGst(root->right);
        val += root->val;
        root->val = val;
        if (root->left)
            bstToGst(root->left);
        return root;
    }
private:
    int val = 0;
};